docs/release-process.md

```diff
@@ -0,0 +1,187 @@
+
+## Support period
+  We release 2 major version per year and we want to support each version at
+least for a year. This means that we will support 2 more versions on top of the
+`latest` version. For example current version `3.x.x` will get at least security
+fixes, until version `6.x.x` is released. Support types:
+  - `latest` version (`major.x.x`) - active minor and patche releases.
+  - `previous` version (`major-1.x.x`) - backport patches and minor fixes
+    only if it improves performance or has high impact.
+  - `life-support` version (`major-2.x.x`)- only backport security fixes.
+  - Anything prior to these is discontinued.
+
+NOTE: We should also collect stats for the active node versions in the network.
+
+## Changelog and release notes
+  CHANGELOG will only create backwards incompatible API change notes and
+actions users need to take to upgrade.  
+ Each version should create new release-note for the version, that will
+contained detailed information of the update. This includes `Pull Requests`
+list (titles and links), CHANGELOG information related to this version.
+
+## Branching strategy
+### General
+  `master` branch is not associated with any released versions and should not
+be used in production. Each released version lives in a separate branch:
+  - `latest` in `major.x` branch.
+  - `previous` in `major-1.x` branch.
+  - `life-support` in `major-2.x` branch.
+
+Minor and patch releases of these versions are released from these branches and
+updates are backported from the `master`. Merges to the released version
+branches must happen via Pull Requests to give other maintainers and reviewers
+time to review and verify what will be released.
+
+Process below will be described as procedures for the situations:
+  - release `next` major version candidate.
+  - release `next` major version. (becomes `latest`)
+  - release `latest` minor and patch versions.
+  - release `previous` minor and patch versions.
+  - release `life-support` minor and patch versions.
+
+Process for the `latest` and `previous` minor/patch releases is the same.
+At the same time only these branches should exist in the repo
+(Example we have just released `v4.0.0`):
+  - `master` - active development.
+  - `v4.x` - just become `latest` version.
+  - `v4.x-proposal` - active backport and PR base.
+  - `v3.x` - just become `previous` version.
+  - `v3.x-proposal` - import minor/patch backport and PR base.
+  - `v2.x` - just become `life-support` version.
+  - `v2` does not have proposal, because it only supports critical fixes, which
+    wont go through standard PR release process.
+
+### master branch and package.json version
+  `master` branch only accumulates changes from PRs and is mostly untested(other
+than PR testing). `next` release candidate will branch off from the master.
+Package version in the `master` branch will always be set to the `next` but
+will have `patch` version set to `99` to signify the unsafe version
+(`latestMajor+1.0.99`).
```

Comment:
Running current master may be v3.1.0 or it may be v4.0.0 depending what changes has landed on master.
That's the main reason master by default is `next major`, even if it does not contain major-breaking changes as of now.

Reply:
Hm, I'm going to push back on this gently one time ;-) No need to bike shed if you wanna keep it this way. One thing I was thinking is that, until `v4.0.0` is officially released, I don't want to see any `v4.x.x` user agents on the network.